common/timedcache/timedcache.go

```diff
@@ -25,18 +25,55 @@ func (te *timedEntry) expired() bool {
 // to expire at exactly the ttl time. The expiration mechanism is 'lazy', and
 // will only remove expired objects at next access.
 type TimedCache struct {
-	ttl   int64     // Time (in seconds) each entry is allowed to live for
-	cache lru.Cache // Underlying size-limited LRU cache
+	ttl   int64
+	cache *lru.Cache
+	lock  sync.RWMutex
+
+	evictedKeys, evictedVals []interface{}
+	onEvictedCB              func(k, v interface{})
 }
 
 // New creates a new cache with a given size and ttl. TTL defines the time in
 // seconds an entry shall live, before being expired.
-func NewTimedCache(size int, ttl int) (*TimedCache, error) {
-	cache, err := lru.New(size)
+func New(size int, ttl int) (*TimedCache, error) {
+	return NewWithEvict(size, ttl, nil)
+}
+
+// NewWithEvict constructs a fixed size cache with the given ttl & eviction
+// callback.
+func NewWithEvict(size int, ttl int, onEvicted func(key, value interface{})) (*TimedCache, error) {
+	tc := &TimedCache{
+		ttl:         int64(ttl),
+		onEvictedCB: onEvicted,
+	}
+	if onEvicted != nil {
+		tc.initEvictBuffers()
+		onEvicted = tc.onEvictedCB
+	}
+	cache, err := lru.NewWithEvict(size, onEvicted)
 	if err != nil {
-		return &TimedCache{}, err
+		return tc, err
 	}
-	return &TimedCache{ttl: int64(ttl), cache: *cache}, nil
+	tc.cache = cache
+	return tc, nil
+}
+
+func (tc *TimedCache) initEvictBuffers() {
+	tc.evictedKeys = make([]interface{}, 0, lru.DefaultEvictedBufferSize)
+	tc.evictedVals = make([]interface{}, 0, lru.DefaultEvictedBufferSize)
+}
+
+// onEvicted save evicted key/val and sent in externally registered callback
+// outside of critical section
+func (tc *TimedCache) onEvicted(k, v interface{}) {
+	tc.evictedKeys = append(tc.evictedKeys, k)
+	tc.evictedVals = append(tc.evictedVals, v)
+}
+
+// calcExpireTime calculates the expiration time given a TTL relative to now.
+func calcExpireTime(ttl int64) int64 {
```

Comment:
Not a big gripe because we do this everywhere else in the codebase, but I'd like to use the builtin `time` package more. They have some nice helpers. This could be a self-contained change.

Reply:
https://github.com/dominant-strategies/go-quai/issues/681